calico/felix/endpoint.py

```diff
@@ -92,8 +92,23 @@ def __init__(self, uuid, mac):
         self.state          = Endpoint.STATE_DISABLED
 
     def remove(self):
-        # Delete a programmed endpoint. Remove the rules only, since the routes
-        # will vanish in due course when the tap interface goes.
+        """
+        Delete a programmed endpoint. Remove the routes, then the rules.
+        """
+        for type in (futils.IPV4, futils.IPV6):
+            try:
+                if devices.tap_exists(self.tap):
+                    ips = devices.list_tap_ips(type, self.tap)
+                    for ip in ips:
+                        devices.del_route(type, ip, self.tap)
+            except:
```

Comment:
Additionally, I think if the tap existed when we started and doesn't exist when we're done we don't need to rethrow. This logic is definitely included in the current code and we want it, but I think we can still get it with just one code block.